Implement descriptions

SimonBaeumer opened the pull request:
This Pull-Requests adds the option to add descriptions into the sh scripts, i.e.
```
#!/usr/bin/env bash
#DESCRIPTION: My funny description.
```
And will display something like this:
```
###################
Available commands:

 - build    Builds phar packages
 - clear    Remove build artifacts
 - unit     Execute unit tests

3 script(s) available
```

@JanPietrzyk Humbug fails maybe due dependencies which are used by the project and humbug's phar, if I add humbug to composer directly I have no problems with executing it :/
See https://github.com/humbug/humbug/issues/39

bcremer commented on README.md:
````diff
@@ -185,6 +185,15 @@ header: |
+
+```sh
+#!/usr/bin/env bash
+#DESCRIPTION: My usefull comment.
````
Typo: usefull -> useful

SimonBaeumer commented on src/Listing/DescriptionReader.php:
```diff
@@ -0,0 +1,28 @@
+     * @param array $lines
+     * @return string
+     */
+    public function read(array $lines)
```
self cr, type hint fehlt

JanPietrzyk commented:
Regarding the build: You usage of PHPUnits mock builder seems to be the problem. I removed all calls locally and Humbug works again.

JanPietrzyk commented on src/Listing/ScriptFinder.php:
```diff
@@ -52,7 +58,9 @@ public function getAllScripts(): array
                 }
 
-                $newScript = new Script($path->getPath(), $fileName, $path->getNamespace());
+                $description = $this->scriptDescriptionReader->read(file($path->getPath() . '/' . $fileName));
```
why add a description reader that then is not reading the file content?

JanPietrzyk commented on tests/Integration/Listing/ScriptFinderTest.php:
```diff
@@ -5,36 +5,53 @@
     public function test_script_finder_holds_contract_if_no_paths_present()
     {
-        $finder = new ScriptFinder([]);
+        $finder = new ScriptFinder([], $this->createMock(DescriptionReader::class));
```
the real one could be used as easily here

JanPietrzyk commented on tests/Integration/Listing/ScriptFinderTest.php:
```diff
@@ -5,36 +5,53 @@
-        $finder = new ScriptFinder([new ScriptPath(__DIR__ . '/_scripts'), new ScriptPath(__DIR__ . '/_scripts_with_misc_stuff')]);
+        $finder = new ScriptFinder(
+            [new ScriptPath(__DIR__ . '/_scripts'), new ScriptPath(__DIR__ . '/_scripts_with_misc_stuff')],
+            $this->descriptionReaderMock
```
the real one could be used as easily here

JanPietrzyk commented on tests/Integration/Listing/ScriptFinderTest.php:
```diff
@@ -5,36 +5,53 @@
-        $finder = new ScriptFinder([new ScriptPath(__DIR__ . '/_scripts'), new ScriptPath(__DIR__ . '/_scripts_with_misc_stuff')]);
+        $finder = new ScriptFinder(
+            [new ScriptPath(__DIR__ . '/_scripts'), new ScriptPath(__DIR__ . '/_scripts_with_misc_stuff')],
+            $this->descriptionReaderMock
```
the real one could be used as easily here

JanPietrzyk commented on tests/Integration/Listing/ScriptFinderTest.php:
```diff
@@ -5,36 +5,53 @@
-        $finder = new ScriptFinder([new ScriptPath(__DIR__ . '/_scripts')]);
+        $finder = new ScriptFinder(
+            [new ScriptPath(__DIR__ . '/_scripts')],
+            $this->descriptionReaderMock
```
the real one could be used as easily here

JanPietrzyk commented on tests/Integration/Listing/ScriptFinderTest.php:
```diff
@@ -44,7 +61,10 @@ public function test_script_finder_finds_script_by_name_if_two_directories_are_p
-        $finder = new ScriptFinder([new ScriptPath(__DIR__ . '/_scripts'), new ScriptPath(__DIR__ . '/_scripts_with_misc_stuff', 'biz')]);
+        $finder = new ScriptFinder(
+            [new ScriptPath(__DIR__ . '/_scripts'), new ScriptPath(__DIR__ . '/_scripts_with_misc_stuff', 'biz')],
+            $this->descriptionReaderMock
```
the real one could be used as easily here

JanPietrzyk commented on tests/Integration/Listing/ScriptFinderTest.php:
```diff
@@ -54,8 +74,31 @@ public function test_script_finder_prefixes_script_names_with_namespace_if_prese
-        $finder = new ScriptFinder([new ScriptPath(__DIR__ . '/_scripts_not_valid_directory')]);
+        $finder = new ScriptFinder(
+            [new ScriptPath(__DIR__ . '/_scripts_not_valid_directory')],
+            $this->descriptionReaderMock
```
the real one could be used as easily here

JanPietrzyk commented on tests/Integration/Listing/ScriptFinderTest.php:
```diff
@@ -54,8 +74,31 @@ public function test_script_finder_prefixes_script_names_with_namespace_if_prese
+    /**
+     * @before
+     */
+    public function createEmptyDescriptionReaderMock()
```
please remove. And if you need a method like that I would prefer a setUp method on top of the test file

SimonBaeumer commented:
@JanPietrzyk I removed `createMock` calls and moved the file reading to the `DescriptionReader` class. I didn't have a reason why I implemented the file reading in the `ScriptFinder` ^^
